### ISSUE TITLE
Request for schema.slack.types.rich_text to be supported when making an API call to chat.postMessage

### ISSUE THREAD

**Issue by KetleyL** [slackapi/deno-slack-sdk#197]
Recently spoke to Slack support who pointed me here to request rich_text being supported when making an API call to chat.postMessage. Basically I'm making a custom app where the user opens a form with rich_text fields and then these are submitted to a channel formatted the way the user intended. Had previously tried using the built in send_message function documented here: https://api.slack.com/reference/functions/send_message however we worked out the documentation may be misleading as rich_text isn't supported yet. 

TLDR; could I please request that the schema.slack.types.rich_text becomes supported when making an API call to chat.postMessage

**Comment by filmaj** [comment]
Hello @KetleyL ! Thanks for writing in. Let me try to explain the situation regarding rich text inputs and outputs currently in Slack, and hopefully that can help you use it in your coded workflows.

First, wider support for rich_text blocks is actively being worked on across all surfaces and Slack APIs. It's not officially rolled out yet - that is being targeted for closer to the end of this month, so support _right now_ may be a bit wonky or incomplete.

That said, in my own tests, it looks to me like both using `rich_text` as a field in the `OpenForm` Slack Function as well as using `rich_text` layout blocks in messages work fine. After speaking to the Block Kit team, this is just the results of the current development efforts; if you tried using these Block Kit elements in a legacy Slack platform app, you would have worse support for it.

So: using this in the new automation platform: totally fine. Using it on the legacy platform: full support coming very soon! 

Because this is still under development, the state of the API documentation right now is incomplete. As such, I will leave this issue open until the work is complete and rolled out to production, with docs, at which point I'll come back here and update and close this issue.

As an example of how to use it, I made a small change in a branch to the deno-request-time-off sample ([branch `richer-text` here](https://github.com/slack-samples/deno-request-time-off/compare/richer-text)) to show off how to use `rich_text` as a parameter. I hope it is helpful!

**Comment by KetleyL** [comment]
Hi Filmaj, thank you for this! Good to know that wider support is imminent and thank you for the examples also, looking forward to using this in the field. Will await the roll out and the update, cheers again!

**Comment by filmaj** [comment]
Hey @KetleyL !

Support for rich text has been GA'ed. I do not think any specific changes need to land in this repo to support it; the `blocks` argument of `chat.postMessage` accepts free-form JSON, so you should be able to include [Rich Text Layout blocks](https://api.slack.com/reference/block-kit/blocks#rich_text) as well as [Rich Text Input blocks](https://api.slack.com/reference/block-kit/block-elements#rich_text_input) without issue.

If there's anything else you need here, feel free to re-open this issue.

**Comment by KetleyL** [comment]
Hiya, I have had a look at this, and while I can see that rich_text has been supported for things like the SendMessage build in function, using it in custom functions still seems to be unsupported. 

When using it in with the SendMessage function I have:
FormWorkflow.addStep(Schema.slack.functions.SendMessage, {
   channel_id: EOP.outputs.fields.channel,
   message: `${EOP.outputs.fields.update} test line of text`,
  }); 
 It works as long as its only that output, however if I try to add more to it (like the 'test line of text' or do something like the above it removes the rich_text and posts the message as plain text.
 
 I had tried a custom function which takes the inputs from a form (rich_text) and then posts using the chat.postMessage but this still displays as [object Object] as it did last time, is there something I'm doing wrong? Sorry if my terminology isn't great!

**Comment by filmaj** [comment]
Hey @KetleyL sorry for the late response! I was away on vacation 🌴  but am back now.

I think the experience you are seeing is expected behaviour - let me explain why. Rich text is represented by a nested JSON, and can be quite complex. Each time you add a separate rich text element, like for example a bold text, or an italic text, or a numbered list, this is represented by a different JSON object. Therefore, you can imagine that for very complex text with different kinds of rich text formatting being used, the resulting JSON object could end up being deeply nested and quite large!

Therefore, adding a string to a JSON object can lead to unexpected behaviours, as you've seen. This is because there is no simple way to add a string to a rich text object - it completely depends on the contents of the rich text!

A custom function should be able to accommodate your situation, but I stress that it will once again depend on the structure of the rich text JSON object. You will have to carefully inspect the structure and nesting of the returned rich text JSON object, and if you want to add additional text to it, you will have to insert into the rich text JSON appropriately in a way that conforms to the rich text model.

> I had tried a custom function which takes the inputs from a form (rich_text) and then posts using the chat.postMessage but this still displays as [object Object] as it did last time, is there something I'm doing wrong?

This should work as long as you are using the `chat.postMessage` API with [the `blocks` parameter](https://api.slack.com/methods/chat.postMessage#arg_blocks). If you try to pass a rich text JSON to the `chat.postMessage` API's `text` parameter, then that would not work, as `text` expects a plain text string. Also, make sure that what you pass to the `blocks` parameter is a [Rich Text layout Block](https://api.slack.com/reference/block-kit/blocks#rich_text) - sort of like a rich text container. Within the layout block, you will have an `elements` array that will contain the actual rich text JSON. The [example here](https://api.slack.com/reference/block-kit/blocks#rich_text_examples) may be helpful. I also recommend using Block Kit Builder for quickly testing out how a rich text block visually looks like right from your browser - [here is a link to the example I linked above viewed in Block Kit Builder](https://app.slack.com/block-kit-builder/T029V6468RL#%7B%22blocks%22:%5B%7B%22type%22:%22rich_text%22,%22elements%22:%5B%7B%22type%22:%22rich_text_section%22,%22elements%22:%5B%7B%22type%22:%22text%22,%22text%22:%22Hello%20there,%20I%20am%20a%20basic%20rich%20text%20block!%22%7D%5D%7D%5D%7D,%7B%22type%22:%22rich_text%22,%22elements%22:%5B%7B%22type%22:%22rich_text_section%22,%22elements%22:%5B%7B%22type%22:%22text%22,%22text%22:%22Hello%20there,%20%22%7D,%7B%22type%22:%22text%22,%22text%22:%22I%20am%20a%20bold%20rich%20text%20block!%22,%22style%22:%7B%22bold%22:true%7D%7D%5D%7D%5D%7D,%7B%22type%22:%22rich_text%22,%22elements%22:%5B%7B%22type%22:%22rich_text_section%22,%22elements%22:%5B%7B%22type%22:%22text%22,%22text%22:%22Hello%20there,%20%22%7D,%7B%22type%22:%22text%22,%22text%22:%22I%20am%20an%20italic%20rich%20text%20block!%22,%22style%22:%7B%22italic%22:true%7D%7D%5D%7D%5D%7D,%7B%22type%22:%22rich_text%22,%22elements%22:%5B%7B%22type%22:%22rich_text_section%22,%22elements%22:%5B%7B%22type%22:%22text%22,%22text%22:%22Hello%20there,%20%22%7D,%7B%22type%22:%22text%22,%22text%22:%22I%20am%20a%20strikethrough%20rich%20text%20block!%22,%22style%22:%7B%22strike%22:true%7D%7D%5D%7D%5D%7D%5D%7D). Finally, I would inspect the output you receive from the OpenForm step's rich text field to see if it is wrapped in a rich text layout block or not.

Hope that's helpful! If you need more help with this issue, feel free to re-open this issue and/or at-mention me.
